identity-wallet/src/state/credentials/reducers/handle_oid4vp_authorization_request.rs

```diff
@@ -124,44 +123,34 @@ pub async fn handle_oid4vp_authorization_request(mut state: AppState, action: Ac
         }
         info!("response successfully sent");
 
-        let connection_time = chrono::Utc::now().to_rfc3339();
-
         let (client_name, _logo_uri, connection_url) =
             get_oid4vp_client_name_and_logo_uri(&oid4vp_authorization_request)
                 .map_err(|_| MissingAuthorizationRequestParameterError("connection_url"))?;
 
-        let mut connections = state.connections.clone();
-        let result = connections
-            .iter_mut()
-            .find(|connection| connection.url == connection_url && connection.client_name == client_name)
-            .map(|connection| {
-                connection.last_interacted = connection_time.clone();
-            });
-
-        let connection_id = Connection::create_connection_id(&client_name);
-
-        if result.is_none() {
-            connections.push(Connection {
-                id: connection_id.to_string(),
-                client_name: client_name.to_string(),
-                url: connection_url,
-                verified: false,
-                first_interacted: connection_time.clone(),
-                last_interacted: connection_time.clone(),
-            })
-        };
-
-        state.connections = connections;
+        let previously_connected = state.connections.contains(connection_url.as_str(), &client_name);
+        let mut connections = state.connections;
```

Comment:
This is actually a good example of why I like the `state` to be immutable by default :) 

My preferred impementation is to go back to how the reducers where initially designed, so in this reducer it would be like this:
```rust
       let previously_connected = state.connections.contains(connection_url.as_str(), &client_name);
        let mut connections = state.connections; // <-- I only want this to be mutable.
        let connection = connections.insert_or_update(&connection_url, &client_name);

        let mut history = state.history; // <-- as well as this
        // It is very easy to see what data I want to change in this reducer. All other fields have no business to be mutable so always safer to keep them immutable by default

        // History
        if !previously_connected {
            // Only add a `ConnectionAdded` event if the connection was not previously connected.
            history.push(HistoryEvent {
                connection_name: connection.name.clone(),
                event_type: EventType::ConnectionAdded,
                connection_id: connection.id.clone(),
                date: connection.last_interacted.clone(),
                credentials: vec![],
            });
        }
        history.push(HistoryEvent {
            connection_name: connection.name.clone(),
            event_type: EventType::CredentialsShared,
            connection_id: connection.id.clone(),
            date: connection.last_interacted.clone(),
            credentials: history_credentials,
        });

        drop(state_guard);
        // Now down here, again it is very easy to see what parts of the state will be updated.
        return Ok(AppState {
            connections,
            history,
            current_user_prompt: Some(CurrentUserPrompt::Redirect {
                target: "me".to_string(),
            }),
            ..state
        });
```

Now the added benefit of using this approach, is that the rust compiler gives me a nice compile error when I forgot to add one of the data variables that I have been mutating back to the `AppState` that will be returned:
```rust
        return Ok(AppState {
            // connections, <-- forgot to put this back into the state
            history,
            current_user_prompt: Some(CurrentUserPrompt::Redirect {
                target: "me".to_string(),
            }),
            ..state
        });
```
will result in this error:
```
131 |           let mut connections = state.connections;
    |                                 ----------------- value moved here
...
156 |           return Ok(AppState {
    |  ___________________^
157 | |             // connections,
158 | |             history,
159 | |             current_user_prompt: Some(CurrentUserPrompt::Redirect {
...   |
162 | |             ..state
163 | |         });
    | |_________^ value used here after move
```

So Rust is trying to take ownership of everything in the current state like this `..state`, but it is failing because I explicitly moved ownership over that field in line 131.

This way I will always be reminded to make sure to add all mutated data in the updated `AppState` that will be returned :)

Reply:
Yes agreed